packages/transaction-controller/src/TransactionController.ts

```diff
@@ -901,46 +923,55 @@ export class TransactionController extends BaseController<
   private async processApproval(
     transactionMeta: TransactionMeta,
     {
+      isExisting = false,
       requireApproval,
       shouldShowRequest = true,
     }: {
+      isExisting?: boolean;
       requireApproval?: boolean | undefined;
       shouldShowRequest?: boolean;
     },
   ): Promise<string> {
     const transactionId = transactionMeta.id;
     let resultCallbacks: AcceptResultCallbacks | undefined;
+    const { meta, isCompleted } = this.isTransactionCompleted(transactionId);
+    const finishedPromise = isCompleted
+      ? Promise.resolve(meta)
+      : this._waitForTransactionFinished(transactionId);
+
+    if (!isExisting && !isCompleted) {
+      try {
+        if (requireApproval !== false) {
+          const acceptResult = await this.requestApproval(transactionMeta, {
+            shouldShowRequest,
+          });
+          resultCallbacks = acceptResult.resultCallbacks;
+        }
 
-    try {
-      if (requireApproval !== false) {
-        const acceptResult = await this.requestApproval(transactionMeta, {
-          shouldShowRequest,
-        });
-        resultCallbacks = acceptResult.resultCallbacks;
-      }
-
-      const { meta, isCompleted } = this.isTransactionCompleted(transactionId);
-
-      if (meta && !isCompleted) {
-        await this.approveTransaction(transactionId);
-      }
-    } catch (error: any) {
-      const { meta, isCompleted } = this.isTransactionCompleted(transactionId);
-
-      if (meta && !isCompleted) {
-        if (error.code === errorCodes.provider.userRejectedRequest) {
-          this.cancelTransaction(transactionId);
+        const { isCompleted: isTxCompleted } =
+          this.isTransactionCompleted(transactionId);
 
-          throw ethErrors.provider.userRejectedRequest(
-            'User rejected the transaction',
-          );
-        } else {
-          this.failTransaction(meta, error);
+        if (!isTxCompleted) {
+          await this.approveTransaction(transactionId);
+        }
+      } catch (error: any) {
+        const { isCompleted: isTxCompleted } =
```

Comment:
I appreciate we always have a reference to `meta` now from the start of the method, but just to be sure, in case some error or code path has removed our original metadata, should we keep the local reference here and check that instead?